perllib/FixMyStreet/App/Form/Waste/Garden.pm

```diff
@@ -10,14 +10,25 @@ has_field is_staff => ( type => 'Hidden' );
 sub details_update_fields {
     my $form = shift;
     my $data = $form->saved_data;
-
+    my $c = $form->{c};
+
+    my $bin_count = 0;
+    my $wanted;
+    if ( $c->get_param('bins_wanted') ) {
+        $wanted = $c->get_param('bins_wanted');
+        $bin_count = $wanted;
+    } else {
+        $data->{existing_number} = 0 if $data->{existing} eq 'no';
+        $bin_count = $data->{existing_number} || 0;
+    }
     my $existing = $data->{existing_number} || 0;
-    $existing = 0 if $data->{existing} eq 'no';
-    my $cost = $existing == 0 ? 0 : $form->{c}->cobrand->garden_waste_cost($data->{existing_number});
+    $wanted ||= $existing;
+
+    my $cost = $bin_count == 0 ? 0 : $form->{c}->cobrand->garden_waste_cost($bin_count);
     $form->{c}->stash->{payment} = $cost / 100;
     return {
         current_bins => { default => $existing },
-        new_bins => { default => 0 },
+        bins_wanted => { default => $wanted },
     };
 }
```

Comment:
Not sure I understand the logic here. If someone managed to submit with existing=no and existing_number>0, this would set it to 0. But if they'd set bins_wanted as well, it would have left existing_number alone, even though existing=no? And don't wanted/bin_count always end up the same value? Does this have the same outcome:
```
    my $existing = $data->{existing_number} || 0;
    $existing = 0 if $data->{existing} eq 'no';
    my $bin_count = $c->get_param('bins_wanted') || $existing;
    my $cost = $bin_count == 0 ? 0 : $form->{c}->cobrand->garden_waste_cost($bin_count);
    return {
        current_bins => { default => $existing },
        bins_wanted => { default => $bin_count },
    };
```

Reply:
yes, that is much better, thanks.